File: CitadelServer/src/uk/co/gossfunkel/citadelserver/Timer.java
package uk.co.gossfunkel.citadel;

public class Timer {
	
	// -------------------- variables -----------------------------------------
	
	private long lastTime;
	private long now;
	private final double fps = 60.0;
	private final double ns= 1000000000.0 / fps;
	private double delta;
	
	private int frames;
	private int updates;
	
	private long second;
	private int hour;
	private boolean day = true;
	
	// -------------------- constructors --------------------------------------
	
	public Timer() {
		
		lastTime = System.nanoTime();
		now = lastTime;
		delta = 0.0;
		frames = 0;
		updates = 0;
		second = System.currentTimeMillis();
		
	}
	
	// -------------------- methods -------------------------------------------
	
	public void tick() {
		
		lastTime = now;
		now = System.nanoTime();
		delta += (now - lastTime) / ns;
		frames++;
		
	}
	
	public void superTick() {
		updates++;
		delta--;
		lastTime = now;
	}
	
	public void hourTick() {
		hour++;
		if (hour == 12) {
			day = !day;
			hour = 0;
		}
	}
	
	public void resetTick() {
		updates = 0;
		frames = 0;
	}
	
	public int getFPS() {
		return frames;
	}
	
	public String returnFPS() {
		return (updates + " ups, " + frames + " fps");
	}
	
	public void accumulateSecond() {
		// add 1000 to second
		second += 1000;
	}
	
	// -------------------- getters -------------------------------------------
	
	public double getDelta() {
		// return nanosecond difference between now and lastTime
		return delta;
	}
	
	public long getSecond() {
		return second;
	}
	
	public int getHour() {
		return hour;
	}
	
	public boolean getDay() {
		return day;
	}

	// -------------------- setters -------------------------------------------
	
	public void setHour(int h) {
		hour = h;
	}

}

File: RubyCitadelServer/timer.rb
class Timer
  
  # -------------------- constructors --------------------------------------
  
  def initialize
    #TODO more precise time function
    $lastTime = Time.now #nanoTime
    private fps = 60.0
    private ns = 1000000000.0 / fps;
    private day = true
    
    private now = lastTime
    private delta = 0.0
    private frames = 0
    private updates = 0
    private second = Time.now #currentTimeMillis
  end
  
  # -------------------- methods -------------------------------------------
  
  def tick
    lastTime = now
    now = Time.now #nanoTime
    delta += (now - lastTime) / ns
    frames += 1
  end
  
  def superTick
    updates += 1
    delta -= 1
    lastTime = now
  end
  
  def hourTick
    hour += 1
    if hour == 12
      day = !day
      hour = 0
    end
  end
  
  def resetTick
    updates = 0
    frames = 0
  end
  
  def returnFPS
    String.new(updates + " ups, " + frames + " fps")
  end
  
  def accumulateSecond
    second += 1000
  end
  
  # -------------------- getters -------------------------------------------
  
  def getDelta
    delta
  end
  
  def getFPS
    frames
  end
  
  def getSecond
    second
  end
  
  def getHour
    hour
  end
  
  def getDay
    day
  end
  
  # -------------------- setters -------------------------------------------
  
  def setHour (h)
    hour = h
  end
  
end